### ISSUE TITLE
Emailing Quotes/ Invoices

### ISSUE THREAD

**Issue by phantomgannon** [opensourcepos/opensourcepos#2301]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [] Reporting an issue of an unmodified OSPOS installation
- [] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.3.0 - 4f5ad57
- OSPOS git commit hash is: 
- PHP version is:  7.2.12
- MySQL or MariaDB version is: 10.1.37-MariaDB - mariadb.org binary distribution
- OS and version is: Windows 7
- WebServer is: 
- Selected language is: English GB
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: Xampp

### Issue / Bug  / Question / New Feature

I have setup email with sendmail, located at c/sendmail/sendmail.exe. Emailing receipts works fine but emailing quote or invoice doesn't.  I have checked the sendmail error log and there is no error the send button doesn't seem to work. Are the invoice/quote emails configured elsewhere?

![email](https://user-images.githubusercontent.com/45674737/50738516-cd7c8080-11cc-11e9-926f-f88196bbfe6d.jpg)

**Comment by odiea** [comment]
You are correct. Below is the error code.
ERROR - 2019-01-06 09:41:07 --> Severity: error --> Exception: No block-level parent found.  Not good. C:\wamp64\www\my-master\vendor\dompdf\dompdf\src\Positioner\Inline.php 48

**Comment by jekkos** [comment]
Ok can you try to send it from the sale edit screen instead? I used that last week and it worked for me. This case might need investigation

**Comment by jekkos** [comment]
OK for me it works just fine. It looks like there's something specific going on maybe in the invoice setup. also on the demo server it tries to send the invoice. The pdf is readily generated and present in the /tmp folder. It's just not sent as sendmail is not setup.

**Comment by WebShells** [comment]
@jekkos I'll create smtp account to set in demo db so we can test it online directly...

**Comment by daN4cat** [comment]
Send email from demo? I would be careful because it can be exploited to spam people. Considering the experience of someone changing the password and etc. ....

**Comment by WebShells** [comment]
@daN4cat mmmm do you think we could still face issues even after limiting number of sent emails per specific period of time or something alike ? 🤔

**Comment by odiea** [comment]
I think I found the Issue. I was using PHP 7.1.16 and dropped down to PHP 5.6.35 and the issue is not there. Now I have no error and it is getting to my mail relay.

**Comment by jekkos** [comment]
Ok weird. I'm on php 7.2 and no issues.

**Comment by odiea** [comment]
I think that @SteveIreland uses WAMP server also. I see this person is using Xampp. After @SteveIreland gets his sanity back maybe he can look into this for us.

**Comment by odiea** [comment]
Maybe some extension in the Windows version or the Parser is not working. Thanks for the help @SteveIreland

**Comment by jekkos** [comment]
emailing is not configured on the demo server (no sendmail in the docker container) but you could indeed temporarily add an smtp account.

**Comment by jekkos** [comment]
I guess this is configuration related as my emailing setup works fine. So will close unless @phantomgannon has other questions.

**Comment by odiea** [comment]
current master can only get php 5.6 to send envoice pdfs
Screen image
![2019-02-09 11_32_40-window](https://user-images.githubusercontent.com/10962177/52524862-fcb85e80-2c5e-11e9-8d75-1f413f1f6ff2.png)

Pdf Email received. Anyone else have this issue
![2019-02-09 11_33_19-window](https://user-images.githubusercontent.com/10962177/52524874-1fe30e00-2c5f-11e9-98cb-7c7c73570eb1.png)

**Comment by odiea** [comment]
In Quotes and Invoice email I removed html,head and body tags now it sends an invoice and quote. In Invoice_email.css I changed the following and now the Image above displays properly.
#info { width: 100%; margin: 10px 0 60px 0; }

**Comment by odiea** [comment]
Here is an update to this issue. I used dompdf version 8.1 when I complied the program  and I had to do nothing to the invoice or quote email files.  But I do get lots of the following errors which do not prevent the pdf from sending.
Severity: Warning --> A non-numeric value encountered C:\wamp64\www\master5\vendor\dompdf\dompdf\src\Css\Style.php 1326

**Comment by odiea** [comment]
I lied. It quit working after a few tests. Back to square 1 for me.

**Comment by SteveIreland** [comment]
I was testing the latest merge which is intended  to fix a coding error related to generating a PDF.

Unfortunately I was NOT able to generate a PDF version of the attachment.  The problem was NOT related to the merged fix.

Instead, like others before me it was chased down to a failure in the `$dompdf->render();` statement.

**Comment by SteveIreland** [comment]
Okay, after much hacking, testing, cursing, and crying I finally stumbled across the problem.

The issue is that the `dompdf->render()` function doesn't digest non-html characters well.  In essence it does not like to need to deal with the contents of `<pre></pre>` tags at all.

But not only that, it doesn't like special ascii characters in the html stream at all .... even between html tags where they are normally ignored.

To correct this problem is simple enough by replacing a line of code with the following.

`$dompdf->loadHtml(str_replace(array("\n", "\r"), '', $html));`

However, that does break the formatting of the code that is wrapped in the `<pre></pre>` tags.

So it's a two part fix.  Change the formatted customer information to use HTML tags for formatting and then change the dompdf_helper.php as shown above.

**Comment by SteveIreland** [comment]
So the merge to fix the build of the file name issue is good.

I'll put together a pull request as described above in order to address this issue unless someone wants to go a different direction.  But based on the previous post by @odiea,  even an upgrade to a more current version of dompdf (version 8.1) the problem doesn't appear to be cleanly resolved.

**Comment by odiea** [comment]
"dompdf/dompdf": "^0.8.3", is what I changed in composer json and composer lock but the licenses in config still states that it is version .7. So I may have lied to you. With your current change it sends pdfs ok but like you said the company and customer does not show properly on the emailed invoices.

**Comment by odiea** [comment]
I forgot to enable logging with the current version. Still getting the dompdf errors but it sends ok

**Comment by odiea** [comment]
@SteveIreland I added the following nl2br and the <pre  issue looks ok now. See if this works for you.

```
<td id="customer-title">
				<pre><?php if(isset($customer)) { echo nl2br($customer_info); } ?></pre>
			</td>
		</tr>
		<tr>
			<td id="company-title">
				<pre><?php echo $this->config->item('company'); ?></pre>
				<pre><?php echo nl2br($company_info); ?></pre>
			</td>
```

**Comment by SteveIreland** [comment]
The method nl2br() addresses the content but can't be used to fix the nl characters between the tags which also causes problems.

So what I did was simply change the implode() function to use the br tag instead of a nl character (intending to change as little code as possible) and then use a str_replace to zap out any residual nl characters from the html string.

The pull request has been submitted.

But I spotted an issue with tax group formatting that I was sure I had resolved but seems to have returned (or maybe I included it in something that hasn't been merged).

Working on that now.
